Reduce unsafeness in WebGPU module

annevk opened the pull request:
#### 24972cfb4d2c5936126a6d2412287ea86a83b6e1
<pre>
Reduce unsafeness in WebGPU module
<a href="https://bugs.webkit.org/show_bug.cgi?id=295160">https://bugs.webkit.org/show_bug.cgi?id=295160</a>

Reviewed by Mike Wyrzykowski.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Canonical link: <a href="https://commits.webkit.org/296859@main">https://commits.webkit.org/296859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a854904328e8d6af35cff836935bc961f3040b4d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29456 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19886 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115819 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60035 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38044 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83438 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112746 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24015 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98876 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63901 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17024 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93385 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17067 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118611 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36837 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27290 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92444 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37210 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95143 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92267 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23510 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37235 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14979 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32676 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36732 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42202 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36392 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39734 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38101 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/WebGPU/GPUBuffer.h:
```diff
@@ -90,7 +90,7 @@ class GPUBuffer : public RefCountedAndCanMakeWeakPtr<GPUBuffer> {
     GPUBufferMapState m_mapState { GPUBufferMapState::Unmapped };
     std::optional<MapAsyncPromise> m_pendingMapPromise;
-    WeakPtr<GPUDevice, WeakPtrImplWithEventTargetData> m_device;
+    const WeakPtr<GPUDevice, WeakPtrImplWithEventTargetData> m_device;
```
This does not protect anything since the type is WeakPtr and WeakPtr does not keep the object alive. This should not help for the purpose of static analysis and if it does, I believe it would likely be a bug in the static analyzer.

annevk replied:
I know, but I saw people adding these in other PRs so I figured I'd do it as drive-by as well.

cdumez replied:
But what's the purpose?

annevk replied:
I guess I misunderstood. I'll revert that change and those like it.

charliewolfe commented on Source/WebCore/Modules/WebGPU/GPUPresentationContext.cpp:
```diff
@@ -64,12 +64,14 @@ void GPUPresentationContext::unconfigure()
+        if (auto backingCurrentTexture = m_backing->getCurrentTexture(index)) {
+            currentTexture = GPUTexture::create(*backingCurrentTexture, m_textureDescriptor, *m_device.get()).ptr();
+            m_currentTexture = currentTexture.copyRef();
+        }
```
We could just inline `isDestroyed()` and avoid doing all of this

annevk replied:
Thanks for pointing that out, I'll make that change!

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -432,16 +432,16 @@ static void imageBytesForSource(WebGPU::Queue& backing, const auto& sourceDescri
         if (!platformImage)
             return callback({ }, 0, 0);
-        RetainPtr pixelDataCfData = adoptCF(CGDataProviderCopyData(CGImageGetDataProvider(platformImage.get())));
+        RetainPtr pixelDataCfData = adoptCF(CGDataProviderCopyData(RetainPtr { CGImageGetDataProvider(platformImage.get()) }.get()));
```
does `RetainPtr` here extend the lifetime of the result returned from `GImageGetDataProvider(platformImage.get())`? Wouldn't `RetainPtr::~RetainPtr` run prior to the invocation of `CGDataProviderCopyData`?

cdumez replied:
> does `RetainPtr` here extend the lifetime of the result returned from `GImageGetDataProvider(platformImage.get())`?

Yes.

> Wouldn't `RetainPtr::~RetainPtr` run prior to the invocation of `CGDataProviderCopyData`?

I do not think so?

mwyrzykowski replied:
oh yes, dtor runs after the function calls completes :), I parsed it incorrectly

annevk committed:
Reduce unsafeness in WebGPU module
https://bugs.webkit.org/show_bug.cgi?id=295160

Reviewed by Mike Wyrzykowski.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Canonical link: https://commits.webkit.org/296859@main
